[rolling] build failures on macOS + Apple Silicon

## Setup
Hardware: Macbook Pro M1
OS: MacOS Ventura 13.1
Apple clang version: 14.0.0 (clang-1400.0.29.202)
Target: arm64-apple-darwin22.2.0
Thread model: posix
XCode version: 14.2

Commit: latest `rolling`

I encountered a series of errors when building specifically with `rviz_ogre_vendor`.


### Error 1: malformed patch
```
patching file 'PlugIns/OctreeZone/CMakeLists.txt'
patch: **** malformed patch at line 1817: 
make[2]: *** [ogre-v1.12.10-prefix/src/ogre-v1.12.10-stamp/ogre-v1.12.10-patch] Error 2
make[1]: *** [CMakeFiles/ogre-v1.12.10.dir/all] Error 2
make: *** [all] Error 2
---
--- stderr: rviz_ogre_vendor
patch: **** malformed patch at line 1817: 
make[2]: *** [ogre-v1.12.10-prefix/src/ogre-v1.12.10-stamp/ogre-v1.12.10-patch] Error 2
make[1]: *** [CMakeFiles/ogre-v1.12.10.dir/all] Error 2
make: *** [all] Error 2
```

This error was reported here https://github.com/ros2/rviz/issues/925 and a viable fix is offered here https://github.com/ros2/rviz/pull/938 (although targeting `humble`). Once resolved, I encountered another error.

### Error 2: Objective-C
```
--- stderr: rviz_ogre_vendor
error: Objective-C was disabled in PCH file but is currently enabled
1 error generated.
make[5]: *** [OgreMain/CMakeFiles/OgreMain.dir/src/OSX/macUtils.mm.o] Error 1
make[5]: *** Waiting for unfinished jobs....
make[4]: *** [OgreMain/CMakeFiles/OgreMain.dir/all] Error 2
make[3]: *** [all] Error 2
make[2]: *** [ogre-v1.12.10-prefix/src/ogre-v1.12.10-stamp/ogre-v1.12.10-build] Error 2
make[1]: *** [CMakeFiles/ogre-v1.12.10.dir/all] Error 2
make: *** [all] Error 2
make: INTERNAL: Exiting with 11 jobserver tokens available; should be 10!
```
It is caused by a `target_precompile_headers()` cmake call.

The final error after resolving the above is due to the `arm64` architecture of Apple Silicon devices.
### Error 3: Linker
```
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[5]: *** [lib/macosx/Codec_Assimp.1.12.10.dylib] Error 1
make[4]: *** [PlugIns/Assimp/CMakeFiles/Codec_Assimp.dir/all] Error 2
make[4]: *** Waiting for unfinished jobs....
make[3]: *** [all] Error 2
make[2]: *** [ogre-v1.12.10-prefix/src/ogre-v1.12.10-stamp/ogre-v1.12.10-build] Error 2
make[1]: *** [CMakeFiles/ogre-v1.12.10.dir/all] Error 2
make: *** [all] Error 2
make: INTERNAL: Exiting with 11 jobserver tokens available; should be 10!
```

I'll open a PR that address all these issues but would appreciate help with further testing!

--- a/rviz_ogre_vendor/CMakeLists.txt
+++ b/rviz_ogre_vendor/CMakeLists.txt
@@ -191,7 +191,7 @@ macro(build_ogre)
       ${extra_cmake_args}
       -Wno-dev
     PATCH_COMMAND
-      ${Patch_EXECUTABLE} -p1 -N < ${CMAKE_CURRENT_SOURCE_DIR}/pragma-patch.diff &&
+      ${Patch_EXECUTABLE} -p1 -N -l < ${CMAKE_CURRENT_SOURCE_DIR}/pragma-patch.diff &&
       ${Patch_EXECUTABLE} -p1 -N < ${CMAKE_CURRENT_SOURCE_DIR}/relocatable.patch &&
       ${Patch_EXECUTABLE} -p1 -N < ${CMAKE_CURRENT_SOURCE_DIR}/C4267.patch
     COMMAND

--- a/rviz_ogre_vendor/CMakeLists.txt
+++ b/rviz_ogre_vendor/CMakeLists.txt
@@ -125,7 +125,11 @@ macro(build_ogre)
     set(OGRE_CXX_FLAGS "${OGRE_CXX_FLAGS} /w /EHsc")
   elseif(APPLE)
     set(OGRE_CXX_FLAGS "${OGRE_CXX_FLAGS} -std=c++14 -stdlib=libc++ -w")
-    list(APPEND extra_cmake_args "-DCMAKE_OSX_ARCHITECTURES='x86_64'")
+    if (${CMAKE_SYSTEM_PROCESSOR} STREQUAL "arm64")
+      list(APPEND extra_cmake_args "-DCMAKE_OSX_ARCHITECTURES='arm64'")
+    else()
+      list(APPEND extra_cmake_args "-DCMAKE_OSX_ARCHITECTURES='x86_64'")
+    endif()
   else()  # Linux
     set(OGRE_C_FLAGS "${OGRE_C_FLAGS} -w")
     # include Clang -Wno-everything to disable warnings in that build. GCC doesn't mind it

--- a/rviz_ogre_vendor/pragma-patch.diff
+++ b/rviz_ogre_vendor/pragma-patch.diff
@@ -1814,14 +1814,14 @@ diff --git a/PlugIns/OctreeZone/CMakeLists.txt b/PlugIns/OctreeZone/CMakeLists.t
 @@ -23,6 +23,11 @@ generate_export_header(Plugin_OctreeZone
      EXPORT_MACRO_NAME _OgreOctreeZonePluginExport
      EXPORT_FILE_NAME ${PROJECT_BINARY_DIR}/include/OgreOctreeZonePrerequisites.h)
-
+ 
 +if (UNIX)
 +  set_property(TARGET Plugin_OctreeZone APPEND PROPERTY
 +    INSTALL_RPATH ${CMAKE_INSTALL_PREFIX}/${OGRE_LIB_DIRECTORY}/OGRE)
 +endif ()
 +
  ogre_config_framework(Plugin_OctreeZone)
-
+ 
  ogre_config_plugin(Plugin_OctreeZone)
 diff --git a/OgreMain/include/OgreTimer.h b/OgreMain/include/OgreTimer.h
 --- a/OgreMain/include/OgreTimer.h	2019-06-24 16:04:20.000000000 -0700

--- a/rviz_ogre_vendor/pragma-patch.diff
+++ b/rviz_ogre_vendor/pragma-patch.diff
@@ -1845,4 +1845,20 @@ diff --git a/OgreMain/include/OgreTimer.h b/OgreMain/include/OgreTimer.h
 +#pragma warning(pop)
 +#endif
  #endif
-
+diff --git a/CMake/Utils/PrecompiledHeader.cmake b/CMake/Utils/PrecompiledHeader.cmake
+index bfb0059..62b4f6d 100644
+--- a/CMake/Utils/PrecompiledHeader.cmake
++++ b/CMake/Utils/PrecompiledHeader.cmake
+@@ -340,7 +340,11 @@ MACRO(ADD_NATIVE_PRECOMPILED_HEADER _targetName _input)
+     if(NOT OGRE_ENABLE_PRECOMPILED_HEADERS)
+         # do nothing
+     elseif(CMAKE_VERSION GREATER_EQUAL "3.16")
+-        target_precompile_headers(${_targetName} PRIVATE ${_input})
++        if(APPLE)
++            add_compile_options((-Xclang -include -Xclang ${_input}))
++        else()
++            target_precompile_headers(${_targetName} PRIVATE ${_input})
++        endif()
+     elseif(CMAKE_GENERATOR MATCHES "^Visual.*$")
+ 
+         # Auto include the precompile (useful for moc processing, since the use of

